src/does-match.js

```diff
@@ -37,45 +37,108 @@
-      }
-    },
-    options: function(options) {
-      var minWord = options.minWord;
-      if (minWord !== undefined && typeof minWord !== 'number') {
-        throw new Error('`minWord`: expected number');
-      } else if (minWord < 1 || minWord > 10) {
-        throw new Error('`minWord`: expected number between 1 and 10');
-      }
-      if (options.replaceDiacritics !== undefined && typeof options.replaceDiacritics !== 'boolean') {
-        throw new Error('`replaceDiacritics`: expected boolean');
+  var validation = function() {
+    var proto = {
+      text: function(text) {
+        if (isString(text)) {
+          throw new Error('`text`: expected string');
+        }
+      },
+      query: function(query) {
+        if (isString(query)) {
+          throw new Error('`query`: expected string');
+        }
+      },
+      options: function(options) {
+        var minWord = options.minWord;
+        if (minWord !== undefined && typeof minWord !== 'number') {
+          throw new Error('`minWord`: expected number');
+        } else if (minWord < 1 || minWord > 10) {
+          throw new Error('`minWord`: expected number between 1 and 10');
+        }
+        if (options.replaceDiacritics !== undefined && typeof options.replaceDiacritics !== 'boolean') {
+          throw new Error('`replaceDiacritics`: expected boolean');
+        }
       }
+    };
+
+    function isString(src) {
+      return typeof src !== 'string'
     }
+
+    return Object.create(proto);
   };
 
   /*
     match
   */
+  var matching = function() {
+    var proto = {
+      chars: function(charWord, charQuery) {
+        return charWord === charQuery;
+      },
+
+      whole: function(text, query, options) {
+        if (text.indexOf(query)>-1) {
+          return query.length * MULTIPLIERS.MATCH_WHOLE;
+        }
+        return 0;
+      },
+
+      words: function(text, query, options) {
+        var queryWords = query.split(' ').filter(function(word) {
+          return word.length > options.minWord;
+        });
+
+        return queryWords.reduce(function(acc, word) {
+          var didMatch = text.indexOf(word)>-1;
+          return acc + (didMatch ? word.length * MULTIPLIERS.MATCH_WORD : 0);
+        }, 0);
+      },
+
+      lookahead: function(text, query, options) {
+        var relevance = 0;
+
+        query = clearWhitespaces(query);
+        query.split('').forEach(function(charQuery) {
+          var j = text.indexOf(charQuery);
+          var didFindChar = j > -1;
+          var isAdjacent = j === 0;
+
+          if (isAdjacent) {
+            relevance++;
+          }
+
+          if (!didFindChar) {
+            return 0;
+          }
+          // on next iteration, will look in the text hereinafter
+          text = text.substring(parseInt(j) + 1);
+        });
+
+        return relevance;
+      }
+    };
+
+    function clearWhitespaces(str) {
+      return str.replace(/ /g, '');
+    }
+
+    return Object.create(proto);
+  };
+
```

Comment:
You have some benefits from using factories. Imagine you have the following prototype:
```javascript
var animalProto = {
  _secret: 'Lion',
  getName: function getName() {
    return "secret animals is" + this._secret
  }
}
```
User can then override default property without calling the set method, even though we're using `_` to define it as private. Using factories, you avoid this problem, because default property will remain private:
```javascript
var factoryAnimal = function () {
  var secret = 'Lion';

  this.getName = function getName() {
    return "secret animals is" + secret
  }
}
```

Also, using `Object.create` lets you create several objects as you need:

```
var validate1 = validation();
var validate2 = validation();
```

The way your code was built, this would not be possible.

Hope this clarifies. Let me know if you have any questions ;)